modules/core/src/lifecycle/component-state.js

```diff
@@ -33,4 +40,151 @@ export default class ComponentState {
   resetOldProps() {
     this.oldProps = this.component.props;
   }
+
+  //
+  // ASYNC PROP HANDLING
+  //
+
+  // Checks if a prop is overridden
+  hasAsyncProp(propName) {
+    return propName in this.asyncProps;
+  }
+
+  // Returns value of an overriden prop
+  getAsyncProp(propName) {
+    const asyncProp = this.asyncProps[propName];
+    return asyncProp && asyncProp.resolvedValue;
+  }
+
+  isAsyncPropLoading(propName) {
+    const asyncProp = this.asyncProps[propName];
+    return (
+      asyncProp &&
+      asyncProp.pendingLoadCount > 0 &&
+      asyncProp.pendingLoadCount !== asyncProp.resolvedLoadCount
+    );
+  }
+
+  // Updates all async/overridden props (when new props come in)
+  // Checks if urls have changed, starts loading, or removes override
+  setAsyncProps(props) {
+    // TODO - use async props from the layer's prop types
+    for (const propName in ASYNC_PROPS) {
+      // const value = this.layer.props._getOriginalValue(propName)
+      if (propName in props) {
+        const value = props[propName];
+        this._updateAsyncProp(propName, value);
+      }
+    }
+  }
+
+  // Intercept strings (URLs) and Promises and activates loading and prop rewriting
+  _updateAsyncProp(propName, value) {
+    // Makes sure a record exists for this prop
+    this._getAsyncPropData(propName);
+
+    if (value instanceof Promise) {
+      this._loadPromiseProp(propName, value);
+      return;
+    }
+
+    if (typeof value === 'string') {
+      this._loadUrlProp(propName, value);
+      return;
+    }
+
+    // else, normal, non-async value. Just store value for now
+    this._setPropValue(propName, value);
+  }
+
+  // interpret value string as promise and track the "loading"
+  _loadPromiseProp(propName, promise) {
+    if (!this._didAsyncInputValueChange(propName, promise)) {
+      return;
+    }
+
+    this._watchPromise(propName, promise);
+  }
+
+  // interpret value string as url, start and track a new load
+  _loadUrlProp(propName, url) {
+    if (!this._didAsyncInputValueChange(propName, url)) {
+      return;
+    }
+
+    // interpret value string as url and start a new load
+    const {fetch} = this.layer.props;
+    const promise = fetch(url);
+
+    this._watchPromise(propName, promise);
+  }
+
+  // Tracks a promise, sets the prop when loaded, handles load count
+  _watchPromise(propName, promise) {
+    const asyncProp = this.asyncProps[propName];
+    asyncProp.pendingLoadCount++;
+    const loadCount = asyncProp.pendingLoadCount;
```

Comment:
Call it `loadId` instead?

Reply:
>Call it loadId instead?

I do perform a numeric comparison `>=` on L158, not clear that `loadId` captures that, `id` gives connotations of a type that mainly supports equality.